cirq/ops/raw_types.py

```diff
@@ -319,6 +321,17 @@ def _qid_shape_(self) -> Tuple[int, ...]:
         (3, 3) for a 2-qutrit ternary gate.
         """
 
+    def _commutes_with_(self, other: Any, atol: Union[int, float] = 1e-8
+                       ) -> Union[bool, NotImplementedType]:
+        if (qid_shape_protocol.qid_shape(self, None) !=
```

Comment:
Note that `other` may be an operation, or some other object.

Whether or not two gates commute depends on whether or not they are applied to the same qubits, and even the order of those qubits. We need to account for this somehow.

Perhaps we should define a `_commutes_on_qubits_(self, qubits, other)` method. `GateOperation._commutes_` would check for the existence of this method, and forward along if appropriate.

Reply:
Is there a difference between `_commutes_on_qubits_` and `_commutes_with_operation_`?

Reply:
Went with `_commutes_on_qids_`.